### DISCUSSION TITLE
Prompt-Tuning for text-to-image diffusion models

### DISCUSSION THREAD

**Discussion by AHHHZ975** [huggingface/diffusers#9493] [Q&A]
Hi, I have been looking for a simple example/script that shows how I can use the prompt-tuning technique to fine-tune the text encoder of a stable diffusion model. But I could not find any. Could you please introduce me if there is already one? If there is no implementation, I would appreciate any help/available resources for fine-tuning the text encoder with or without the PEFT library. Thanks!

**Comment by asomoza** [comment]
Hi, what is the `prompt-tuning technique` or what do you mean by it?

Also you need to specify the model arch you want to use, there's a whole world of difference between the Clip model that older models use and the T5 that the new ones use.

If you mean the CLIP models, you can learn how to fine tune it in this repo: https://github.com/zer0int/CLIP-fine-tune

**Reply by AHHHZ975** [reply]
> Hi, what is the `prompt-tuning technique` or what do you mean by it?

Thank you for the reply. I meant the method proposed by this paper: https://huggingface.co/papers/2104.08691


> Also you need to specify the model arch you want to use, there's a whole world of difference between the Clip model that older models use and the T5 that the new ones use.

Specifically, I need to fine-tune the stable diffusion model (e.g. "runwayml/stable-diffusion-v1-5" or "CompVis/stable-diffusion-v1-4" or similar ones) using the prompt-tuning method mentioned. So, I thought fine-tuning such a stable diffusion model using this approach would be equal to freezing all the SD model parameters (including the parameters of the UNet, VAE, and the CLIP text encoder) and trying to train and learn new tokens that are prepended to the embedding of the input text prompts that is fed to the CLIP text prompt (the same way as mentioned in the prompt-tuning paper). So, I was wondering if there is already implementation for this or generally Prompt-Tuning for text-to-image diffusion models similar to the implementation of prompt-tuning which exists in the PEFT library for language models.


> If you mean the CLIP models, you can learn how to fine tune it in this repo: https://github.com/zer0int/CLIP-fine-tune

Thank you for sharing this repo. But, I don't think this includes the prompt tuning according to what I just explained, am I right?

**Reply by asomoza** [reply]
if you freeze the unet, the model won't learn anything, it doesn't matter how many new tokens you add, if the unet doesn't know them, it won't know what to do with them.

That paper is for LLMs which are easier to train like that, but for a Diffusion Model depending on what you want, you pretty much always need to train the unet/transformer model too.

As far as a I know, there's isn't code or a successful story of someone training new tokens without training the unet. You can maybe refer to [textual inversions](https://arxiv.org/pdf/2208.01618) and [pivotal tuning](https://huggingface.co/blog/sdxl_lora_advanced_script#pivotal-tuning).

Another option you have, is, since the newer models use the T5, maybe you can train one and test the results, but they aren't used for text generation so I don't expect you to get the results you want.

> Thank you for sharing this repo. But, I don't think this includes the prompt tuning according to what I just explained, am I right?

right, that repo finetunes the CLIP models to make better the class label understanding since they're vision transformers not LLMs, they're trained or finetuned with text/image pairs.

**Reply by AHHHZ975** [reply]
> if you freeze the unet, the model won't learn anything, it doesn't matter how many new tokens you add, if the unet doesn't know them, it won't know what to do with them.
> 
> That paper is for LLMs which are easier to train like that, but for a Diffusion Model depending on what you want, you pretty much always need to train the unet/transformer model too.

I wonder why the model won't learn anything. Because my intuition is that by freezing the unet model, we are seeking the domain of those prompts that yield the best output images ("best" in terms of the objective during the training) without changing the unet and without forgetting previous tasks. So, in this case, the fine-tuning process should guide us to a specific domain of text prompts that is more aligned with the fine-tuning objective. Am I right? Can you also please explain why you think freezing the unet won't work?


> As far as a I know, there's isn't code or a successful story of someone training new tokens without training the unet. You can maybe refer to [textual inversions](https://arxiv.org/pdf/2208.01618) and [pivotal tuning](https://huggingface.co/blog/sdxl_lora_advanced_script#pivotal-tuning).

Thanks for sharing these two papers.


> Another option you have, is, since the newer models use the T5, maybe you can train one and test the results, but they aren't used for text generation so I don't expect you to get the results you want.

Yes, I have already experimented with the T5 with prompt-tuning in the peft library but do you think it is possible to replace the CLIP in the SD pipeline with the T5 which is already supported in the Peft library? In that case, one can learn to fine-tune T5 using the prompt-tuning in peft and use it as a text encoder for the tex-to-image SD pipeline.

**Reply by AHHHZ975** [reply]
> > if you freeze the unet, the model won't learn anything, it doesn't matter how many new tokens you add, if the unet doesn't know them, it won't know what to do with them.
> > That paper is for LLMs which are easier to train like that, but for a Diffusion Model depending on what you want, you pretty much always need to train the unet/transformer model too.
> 
> I wonder why the model won't learn anything. Because my intuition is that by freezing the unet model, we are seeking the domain of those prompts that yield the best output images ("best" in terms of the objective during the training) without changing the unet and without forgetting previous tasks. So, in this case, the fine-tuning process should guide us to a specific domain of text prompts that is more aligned with the fine-tuning objective. Am I right? Can you also please explain why you think freezing the unet won't work?
> 
> > As far as a I know, there's isn't code or a successful story of someone training new tokens without training the unet. You can maybe refer to [textual inversions](https://arxiv.org/pdf/2208.01618) and [pivotal tuning](https://huggingface.co/blog/sdxl_lora_advanced_script#pivotal-tuning).
> 
> Thanks for sharing these two papers. I had actually faced the textual inversion before.
> 
> > Another option you have, is, since the newer models use the T5, maybe you can train one and test the results, but they aren't used for text generation so I don't expect you to get the results you want.
> 
> Yes, I have already experimented with the T5 with prompt-tuning in the peft library but do you think it is possible to replace the CLIP in the SD pipeline with the T5 which is already supported in the Peft library? In that case, one can learn to fine-tune T5 using the prompt-tuning in peft and use it as a text encoder for the tex-to-image SD pipeline.

**Reply by AHHHZ975** [reply]
> As far as a I know, there's isn't code or a successful story of someone training new tokens without training the unet. You can maybe refer to [textual inversions](https://arxiv.org/pdf/2208.01618) and [pivotal tuning](https://huggingface.co/blog/sdxl_lora_advanced_script#pivotal-tuning).

Also, as far as I understood, the Textual Inversion step also freezes the text-to-image SD model and only learns the embedding of the text.  But, the most important difference between the textual inversion and the original prompt tuning paper is that we need to provide their model with several sample images and then, they do an inversion step on the text prompt corresponding to the sample images (not images) and then try to learn those noisy inversion of texts during the denoising steps using a reconstruction loss. But, in the original prompt-tuning paper, they only consider several fixed-size learnable token vectors append them to the embedding of the input text prompt, and then freeze all the parameters except for the newly added tokens. Am I right? So, can we conclude that only learning the text embedding space while keeping the unet frozen can give us some results?

**Reply by asomoza** [reply]
> I wonder why the model won't learn anything. Because my intuition is that by freezing the unet model, we are seeking the domain of those prompts that yield the best output images ("best" in terms of the objective during the training) without changing the unet and without forgetting previous tasks. So, in this case, the fine-tuning process should guide us to a specific domain of text prompts that is more aligned with the fine-tuning objective. Am I right? Can you also please explain why you think freezing the unet won't work?

That's because I don't really know what are you training for, for example what task are we referring here. 

When you say `the fine-tuning process should guide us to a specific domain of text prompts that is more aligned with the fine-tuning objective`, there won't be any problem if the fine-tuning objective is something the unet knows, for example `generating images` in a domain it was trained but if it's something the unet still doesn't know e.g.: medical pathologies (x-rays) or  [instruct pix to pix](https://www.timothybrooks.com/instruct-pix2pix) it won't work.

These are my assumptions though because I haven't trained something like this and haven't seen someone else do it.

> do you think it is possible to replace the CLIP in the SD pipeline with the T5 which is already supported in the Peft library

It's possible, there was a project called [ELLA](https://github.com/TencentQQGYLab/ELLA) that did this, they released the code and weights for SD 1.5. Maybe that will work for what you need.

**Reply by asomoza** [reply]
> Am I right? So, can we conclude that only learning the text embedding space while keeping the unet frozen can give us some results?

Yes, but in my experience textual inversions only work with concepts the model already knows, it doesn't make it learn new concepts and that's why they aren't as popular as LoRAs. Textual Inversion these days are mostly used to replace negative prompts or to guide the model to, for example, generate better hands (which in my experience never worked).

If you want to do something similar, in theory, it should also work.

**Reply by AHHHZ975** [reply]
> That's because I don't really know what are you training for, for example what task are we referring here.
> 
> When you say `the fine-tuning process should guide us to a specific domain of text prompts that is more aligned with the fine-tuning objective`, there won't be any problem if the fine-tuning objective is something the unet knows, for example `generating images` in a domain it was trained but if it's something the unet still doesn't know e.g.: medical pathologies (x-rays) or [instruct pix to pix](https://www.timothybrooks.com/instruct-pix2pix) it won't work.
> 
> These are my assumptions though because I haven't trained something like this and haven't seen someone else do it.

Oh right, that totally makes sense to me now. But yes, my goal is to just search within the domain that the model was trained on before.

> It's possible, there was a project called [ELLA](https://github.com/TencentQQGYLab/ELLA) that did this, they released the code and weights for SD 1.5. Maybe that will work for what you need.

Interesting! Thanks for sharing it.

> Yes, but in my experience textual inversions only work with concepts the model already knows, it doesn't make it learn new concepts and that's why they aren't as popular as LoRAs. Textual Inversion these days are mostly used to replace negative prompts or to guide the model to, for example, generate better hands (which in my experience never worked).
> 
> If you want to do something similar, in theory, it should also work.

Yes, it's true but I'd like to know, according to your experience, what general steps should be taken to generally get this done. , I am thinking that if methods like the prompt-tuning under the umbrella of the PEFT are working well in the area of LLMs, then I guess they should also work in the domain of text-to-image diffusion models (and generally computer vision) with some engineering tricks/implementations + solid developed theory. Am I right? I know that my question might be too general but I'd be happy to know your opinion on this and also what is the reason that methods like Textual Inversion do not work well.

**Reply by asomoza** [reply]
> what is the reason that methods like Textual Inversion do not work well

Is not that Textual Inversion(TI) do not work well but more that the utility of them was superseded by LoRAs and better models, with SD 1.5 TI were used a lot, but when SDXL came, they lost the usefulness because SDXL mostly doesn't need negative prompts and you can't make the model do better generations with TI, you need to fine tune the model with millions of images.

 Even more, with Dreambooth you can teach it to generate any kind of concept or subject, contrary to TI that can't teach it something new like for example how to generate a photo of yourself. Finally training LoRAs is really fast and cheap right now, even Flux which is a 12B model can be trained in a couple of minutes (online).

TI still has the utility with pivotal tuning, which is to prevent catastrophic forgetting of the previous concept, useful for people that use multiple loras of characters/persons.

> if methods like the prompt-tuning under the umbrella of the PEFT are working well in the area of LLMs, then I guess they should also work in the domain of text-to-image diffusion models with some engineering tricks/implementations + solid developed theory ... what general steps should be taken to generally get this done

I still fail to understand what is your objective here, maybe @sayakpaul or @BenjaminBossan can answer better your question.

**Reply by BenjaminBossan** [reply]
See discussion here: https://github.com/huggingface/peft/issues/2085
